src/thor/bidirectional_astar.cc

```diff
@@ -916,10 +918,167 @@ void BidirectionalAStar::SetDestination(GraphReader& graphreader, const valhalla
   }
 }
 
+/**
+ * Recover shortcut edge and recost it
+ * @param  graphreader  Graph tile reader.
+ * @param  shortcut_id  Edge id of the shortcut to be recosted
+ * @param  costing      Costing function
+ * @param  transition_cost  Transition cost for the shortcut edge
+ * @param  previous_cost    Total cost from the previous edge
+ * TODO: add time information
+ */
+std::vector<PathInfo> recost_shortcut_forward(GraphReader& graphreader,
+                                              const GraphId& shortcut_id,
+                                              const DynamicCost& costing,
+                                              Cost transition_cost,
+                                              Cost previous_cost) {
+  auto edges = graphreader.RecoverShortcut(shortcut_id);
+
+  std::vector<PathInfo> path;
+  // transition cost will not be included for the first edge;
+  // so, add it here
+  previous_cost += transition_cost;
+
+  const auto on_new_label = [&path, &previous_cost, &transition_cost](const EdgeLabel& label) {
+    Cost transition = path.empty() ? transition_cost : label.transition_cost();
+    path.emplace_back(label.mode(), previous_cost + label.cost(), label.edgeid(), 0,
+                      label.restriction_idx(), transition);
+  };
+
+  auto edge_iter = edges.begin();
+  const auto get_next_edge = [&edge_iter, &edges]() -> GraphId {
+    if (edge_iter == edges.end())
+      return {};
+
+    return *edge_iter++;
+  };
+
+  try {
+    sif::recost_forward(graphreader, costing, get_next_edge, on_new_label);
+  } catch (const std::exception& e) {
+    LOG_ERROR(std::string("Failed to recost shortcut edge: ") + e.what());
+    return {};
+  } catch (...) {
+    LOG_ERROR("Failed to recost shortcut edge: unknown exception");
+    return {};
+  }
+
+  return path;
+}
+
+void BidirectionalAStar::FormPathForward(baldr::GraphReader& graphreader,
+                                         uint32_t last_label_idx,
+                                         std::vector<PathInfo>& path) {
+  // we should go from the beginning to the end because of possible recostings;
+  // so, at first we should collect edge label indexes
+  std::stack<uint32_t> forward_idxs;
+  // Recover forward labels using predecessor indexes
+  for (auto edgelabel_index = last_label_idx; edgelabel_index != kInvalidLabel;
+       edgelabel_index = edgelabels_forward_[edgelabel_index].predecessor()) {
+    forward_idxs.push(edgelabel_index);
+  }
```

Comment:
why make this change? its a bunch of extra allocation and a whole extra loop over the path. what does this give us other than a slightly clearer `while` loop below?

Reply:
unless there is a clear reason to do this, i would highly suggest we revert this to using the label index and labelset directly in the while loop as was done before

Reply:
oh i see in the comment hints at it. you are saying that we have to start at the beginning to get the correct cost? i'm not sure its actually true though. i wonder what kind of penalty this adds. i think we can work around it by looking one label ahead and using that label as the "previous cost" for the recosting of the shortcut to avoid all this allocation

Reply:
let's suppose we go from the end to the beginning (as before), and we already processed some edges: `edge_n, edge_n-1, ...., edge_k` and then meet shortcut and recost the edge - it means that costs for all edges from `n` to `k` will be invalid anymore and we must add difference between shortcut cost and recosted shortcut to each of them.
If we implement this approach directly - it will require approximately `n * (number of shortcuts)` additional operations (change costs) and no additional memory. We can optimize it a little: let's save all cost changes to stack (`~ (number of shortcuts)` additional memory) and apply these changes once on the second pass.